doc/contributing/adding-v8-fast-api.md

```diff
@@ -1,91 +1,397 @@
+throwing errors directly from fast calls. These constraints no longer exist;
+however, a function whose execution cost is far higher than its calling cost is
+unlikely to benefit from having a "fast" variant, so some judgement is required
+when considering whether or not to add a Fast API callback.
+
+## Basics
+
+A Fast API callback must correspond to a conventional ("slow") implementation
+of the same callback. Compare the two conventions:
+
+```cpp
+// Conventional ("slow") implementation
+void IsEven(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  if (!args[0]->IsInt32()) {
+    return THROW_ERR_INVALID_ARG_TYPE(env, "argument must be an integer");
+  }
 
-  ```cpp
-  // Instead of invoking the method as `receiver.internalModuleStat(input)`, the JS land should
-  // invoke it as `internalModuleStat(binding, input)` to make sure the binding is available to
-  // the native land.
-  static int32_t FastInternalModuleStat(
-      Local<Object> unused,
-      Local<Object> recv,
-      const FastOneByteString& input,
-      FastApiCallbackOptions& options) {
-    Environment* env = Environment::GetCurrent(recv->GetCreationContextChecked());
-    // More code
+  int32_t n = args[0]->Int32Value(env->context()).FromJust();
+  bool result = n % 2 == 0;
+  args.GetReturnValue().Set(result);
+}
+
+// Fast implementation
+bool FastIsEven(const int32_t n) {
+  return n % 2 == 0;
+}
+static v8::CFunction fast_is_even(v8::CFunction::Make(FastIsEven));
+```
+
+The main differences between the two call conventions are:
+
+* A conventional call passes its arguments as `v8::Value` objects, via a
+  `v8::FunctionCallbackInfo` object. A Fast API call passes its arguments
+  directly to the C++ function, as native C++ types where possible.
+* A conventional call passes its return value via a `v8::ReturnValue` object,
+  accessible via the `v8::FunctionCallbackInfo` object. A Fast API call returns
+  its value directly from the C++ function, as a native C++ type.
+* A conventional call can pass any number of arguments of any type, which must
+  be validated within the implementation. A Fast API callback will only ever be
+  called in compliance with its function signature, so the `FastIsEven` example
+  above will only ever be called with a single argument of type `int32_t`. Any
+  calls from JavaScript whose arguments do not correspond to a fast callback
+  signature will be directed to the slow path by V8, even if the function is
+  optimized.
+* The fast callback cannot be bound directly. It must first be used to build a
+  `v8::CFunction` handle, which is passed alongside the conventional callback
+  when binding the function.
+
+## Argument and return types
+
+The following are valid argument types in Fast API callback signatures:
+
+* `bool`
+* `int32_t`
+* `uint32_t`
+* `int64_t`
+* `uint64_t`
+* `float`
+* `double`
+* `v8::Local<v8::Value>` (analogous to `any`)
```

Comment:
V8::Object is also accepted

Reply:
https://github.com/nodejs/node/pull/58999#discussion_r2193602025